ugvc/pipelines/correct_genotypes_by_imputation.py

```diff
@@ -0,0 +1,495 @@
+        required=False,
+        type=str,
+    )
+    ap_var.add_argument(
+        "--temp_dir",
+        help="Path to temporary dir",
+        required=False,
+        type=str,
+        default="/tmp",
+    )
+    ap_var.add_argument(
+        "--threads_for_contig", help="Number of threads to use when looping across contigs", type=int, default=4
+    )
+    ap_var.add_argument(
+        "--threads_beagle", help="Number of threads when running Beagle per contig", type=int, default=1
+    )
+    ap_var.add_argument(
+        "--verbosity",
+        help="Verbosity: ERROR, WARNING, INFO, DEBUG",
+        required=False,
+        default="INFO",
+    )
+    ap_var.add_argument("--add_imp_effect", help="Is the input a result of mutect", action="store_true")
+
+    return ap_var
+
+
+def _command_to_subset_vcf(chrom: str, vcf: str, outfile: str) -> str:
+    cmd = f"""bcftools view -O z -o {outfile} {vcf} {chrom} && \
+    bcftools index -t {outfile}
+    """
+    return cmd
+
+
+def _command_to_filter_gq_vcf(vcf: str, outfile: str) -> str:
+    cmd = f"""bcftools view -f PASS {vcf} | \
+    bcftools filter -i 'QUAL>20 && FORMAT/GQ[0]>20' | \
+    bcftools view -O z -o {outfile} && \
+    bcftools index -t {outfile}
+    """
+    return cmd
+
+
+def _command_to_run_beagle(vcf: str, ref: str, gmap: str, outfile: str, nthreads: int = 1) -> str:
+    outprefix = outfile.replace(".vcf.gz", "")
+    cmd = f"""beagle -Xmx7g \
+    gt={vcf} \
+    ref={ref} \
+    map={gmap} \
+    out={outprefix} \
+    nthreads={nthreads} window=100 && \
+    bcftools index -t {outprefix}.vcf.gz
+    """
+    return cmd
+
+
+def _command_to_filter_and_collpase_beagle(vcf: str, outfile: str) -> str:
+    cmd = f"""bcftools view -i 'GT="alt"' {vcf} | \
+    grep -v END | \
+    bcftools norm --multiallelics + -o {outfile} -O z && \
+    bcftools index -t {outfile}
+    """
+    return cmd
+
+
+def _command_to_annotate_with_beagle(vcf: str, beagle_collapsed_vcf: str, outfile: str) -> str:
+    cmd = f"""bcftools annotate --annotations {beagle_collapsed_vcf} \
+    --columns INFO/IMP,INFO/DR2,FORMAT/DS -o {outfile} -O z  {vcf} && \
+    bcftools index -t {outfile}
+    """
+    return cmd
+
+
+def genotype_ordering(num_alt: int) -> np.ndarray:
+    # Returns a numpy array with the order of the genotypes (based on the section Genotype Ordering in the VCF spec).
+    # Each row is a genotype, and the columns are the alleles.
+    gr_ar = np.full([int((num_alt + 2) * (num_alt + 1) / 2), 2], fill_value=-1, dtype=int)
+    if num_alt == 1:
+        gr_ar = np.array([[0, 0], [0, 1], [1, 1]])
+    elif num_alt == 2:
+        gr_ar = np.array([[0, 0], [0, 1], [1, 1], [0, 2], [1, 2], [2, 2]])
+    else:
+        i = 0
+        for a2 in range(num_alt + 1):
+            for a1 in range(a2 + 1):
+                gr_ar[i, 0] = a1
+                gr_ar[i, 1] = a2
+                i += 1
+    return gr_ar
+
+
+def phred(p: tuple[int] | np.ndarray) -> np.ndarray:
```

Comment:
I think this should have been implemented in ugvc already and if not should belong to more general code